virtcontainers/cgroups.go

```diff
@@ -0,0 +1,111 @@
+// +build linux
+// Copyright (c) 2018 Huawei Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package virtcontainers
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/kata-containers/runtime/virtcontainers/pkg/annotations"
+	"github.com/opencontainers/runc/libcontainer/cgroups"
+	"github.com/opencontainers/runc/libcontainer/cgroups/fs"
+	"github.com/opencontainers/runc/libcontainer/configs"
+	"github.com/opencontainers/runc/libcontainer/specconv"
+	"github.com/opencontainers/runtime-spec/specs-go"
+)
+
+// cgroupsManager maintains cgroups implementations and config
+type cgroupsManager struct {
+	libcontainerManager cgroups.Manager
+	libcontainerConfig  *configs.Config
+}
+
+// newManager setup cgroup manager for sandbox
+func (cgm *cgroupsManager) newManager(s *Sandbox) error {
+	ociConfigStr, err := s.Annotations(annotations.ConfigJSONKey)
+	if err != nil {
+		return err
+	}
+
+	var ociSpec specs.Spec
+	if err := json.Unmarshal([]byte(ociConfigStr), &ociSpec); err != nil {
+		return err
+	}
+
+	cgm.libcontainerConfig, err = specconv.CreateLibcontainerConfig(&specconv.CreateOpts{
+		CgroupName:       s.id,
+		UseSystemdCgroup: false,
+		NoPivotRoot:      false,
+		NoNewKeyring:     false,
+		Spec:             &ociSpec,
+	})
+
+	if err != nil {
+		return fmt.Errorf("create container config for %s failed with %s", s.id, err)
+	}
+
+	state, _ := s.storage.fetchSandboxState(s.id)
+	if state.CgroupPaths == nil {
```

Comment:
I don't really understand this code. Why can't it be just...

```
state, err := s.storage.fetchSandboxState(s.id)
if err != nil {
    return err
}

cgm.libcontainerManager = &fs.Manager{
    Cgroups:    cgm.libcontainerConfig.Cgroups,
    Paths:      state.CgroupPaths,
}
```

Maybe a comment in the code would help here.